weave/trace/context/weave_client_context.py

```diff
@@ -43,3 +45,36 @@ def require_weave_client() -> WeaveClient:
     if (client := get_weave_client()) is None:
         raise WeaveInitError("You must call `weave.init(<project_name>)` first")
     return client
+
+
+@contextmanager
+def with_weave_client(
+    entity: str | None, project: str | None, required: bool = True
+) -> Generator[WeaveClient | None, None, None]:
+    """Context manager that returns a WeaveClient for a given entity and project.
+
+    Restores any active WeaveClient after the context manager exits.
+
+    Args:
+        entity: Entity name or None to use the current entity if there is one
+        project: Project name or None to use the current project if there is one
+        required: If True, raises WeaveInitError when no client exists and entity/project are None
+
+    Yields:
+        A WeaveClient to use for the duration of the context manager or None if no client exists
+        and required=False
+    """
+    current_client = get_weave_client()
+    if entity is None and project is None:
+        if required and current_client is None:
+            raise WeaveInitError("You must call `weave.init(<project_name>)` first")
+        yield current_client
+    else:
+        from weave.trace.weave_init import init_weave
+
+        scoped_name = f"{entity}/{project}"
+        try:
+            initialized_client = init_weave(scoped_name, ensure_project_exists=False)
```

Comment:
what happens if you have a current client AND you define an entity/project? Will that leave behind a global client different from the one already running?

Reply:
I don't believe so, the finally clause should restore the global client back to what it was.